### ISSUE TITLE
Object Storage

### ISSUE THREAD

**Issue by perriea** [VictoriaMetrics/VictoriaMetrics#38]
Is it expected that the project will support object storage like Thanos or Cortex ?
Object storage like `S3` or `GCS`.

**Comment by valyala** [comment]
Thanks for the idea, @perriea ! Could you provide more context where this may be necessary?

Probably it is possible to point VictoriaMetrics (via `-storageDataPath`) to a directory mounted from `S3` or `GCS` via [FUSE](https://en.wikipedia.org/wiki/Filesystem_in_Userspace) wrappers such as [gcsfuse](https://github.com/GoogleCloudPlatform/gcsfuse) or [s3fs-fuse](https://github.com/s3fs-fuse/s3fs-fuse). Could you try such setup and report results here?

**Comment by MaxDiOrio** [comment]
Thanos allows storing of metric data directly in S3 or GCS using Minio.

I'd love to see snapshots be able to be written directly to S3 or GCS.  One step backup and truncation of data.

**Comment by perriea** [comment]
Yes, exactly @MaxDiOrio that's what I'm looking for because I'd like to test this project in a Kubernetes context without having to do stateful.
However I will test what I recommended @valyala although I know very well that the system fuse is not crazy in terms of performance.

**Comment by valyala** [comment]
> I'd love to see snapshots be able to be written directly to S3 or GCS. One step backup and truncation of data.

Unfortunately this isn't easy as it sounds. Automatic uploading of snapshots to GCS and S3 may be interrupted due to various failures:
- network connectivity issues
- intermittent object store maintenance / outage

The probability of errors increases with the size of snapshot files. VictoriaMetrics may produce multi-TB files in snapshots, which result in very high probability of errors during automatic snapshot uploading. How to handle such errors? There are various approaches for errors handling and uploading strategies with different pros & cons:
- Just retry file upload from the beginning. This may be huge waste of time for multi-TB files.
- Resume the upload from the failed place if this is supported by the object store.
- Split big files in smaller chunks and upload them separately with per-chunk retries on errors. During the recovery merge these chunks into the original files.
- Archive the snapshot into a single file, then split it into moderately sized chunks and upload them in parallel to object store in order to improve upload performance.
- Perform incremental backup, i.e. upload only new files and delete missing files from the object store. This is possible because of [LSM tree used by VictoriaMetrics](https://medium.com/@valyala/how-victoriametrics-makes-instant-snapshots-for-multi-terabyte-time-series-data-e1f3fb0e0282).
- Your approach.

There are other approaches exist. Each approach has certain users, but there is no "silver bullet" approach for snapshot uploading, which would suit all the users. So it is highly unlikely VictoriaMetrics will automatically upload snapshots to object store in the future.

**Comment by valyala** [comment]
> Yes, exactly @MaxDiOrio that's what I'm looking for because I'd like to test this project in a Kubernetes or with (high) scalability context without having to do stateful.

It is better to use persistent volumes backed by replicated durable disks for Kubernetes such as [Google Compute Engine persistent disks](https://cloud.google.com/compute/docs/disks/#pdspecs) - see [these docs](https://cloud.google.com/kubernetes-engine/docs/concepts/persistent-volumes) for details. VictoriaMetrics is optimized for HDD-backed persistent disks, so there is no much sense in using more expensive SSD-backed persistent disks.

**Comment by antoniobeyah** [comment]
Agreed with this request.  Using persistent volumes within Kubernetes in AWS typically defaults to using EBS which is tied to an availability zone which is a challenge when you have a k8s cluster that spans multiple availability zones. You cannot use the same EBS volume across availability zones so it creates issues when your cluster is in an autoscaling group.

**Comment by MaxDiOrio** [comment]
This is what multipart S3 uploads help solve.  I do understand the challenge though.  I'd picture doing an automated monthly backup, which should keep the size more manageable.

**Comment by MaxDiOrio** [comment]
Maybe do a daily snapshot/archive to cloud storage for small file size with a monthly purge?  Definitely check how Thanos is doing this. They are storing and querying directly to cloud storage.  Clearly it's not going to be as performant as local storage but it would be a good option.

It could also be a good option for those of us with on-prem slow storage that we can use with minio as object storage.  Still local and fast but not on our primary, expensive ssd storage.

It would be great to archive off but still be able to query the archived data like Thanos.

If this can be done without the complexity of Thanos that would be amazing.  I spent a full week trying to get the Thanos and sidecar pieces set up but had too many issues.  That drove me to Victoria Metrics as a probable solution.

**Comment by valyala** [comment]
It shouldn't be too hard to store VictoriaMetrics' data in object storage if it supports the following operations:
1) Streaming writing to a file without limits on file size. Resume streaming writing after network error must be supported.
2) Fast concurrent reading of arbitrary parts of the file.

The main issue with object storage is network errors, which would reduce reliability and availability comparing to much more reliable block storage such as [GCE disks](https://cloud.google.com/compute/docs/disks/#pdspecs), [EBS](https://aws.amazon.com/ebs/) or local disks. Currently VictoriaMetrics assumes that it works with reliable storage (aka filesystem), so it reports the first occurred error related to storage and exits instead of trying to recover from such errors (it is unclear how to recover from these errors). When working with object storage, such errors must be handled gracefully, since they are usually expected in object storage world. This is one of the reasons why Thanos has not so good reliability comparing to VictoriaMetrics.

**Comment by valyala** [comment]
There is yet another approach for adding ability to use Object Storage as a store for VictoriaMetrics data: to write a thin layer between file API and Object Storage API, which will split big files into smaller chunks and upload them to Object Storage as separate files. This should alleviate network connectivity issues outlined above and will improve upload retrying speed.

**Comment by perriea** [comment]
It seems like a good idea. How do you see this feature in VictoriaMetrics ? Native functionality ? Plugin ?
I ask because there are different protocols (S3, GCS, Azure Blob storage for the main ones).

**Comment by valyala** [comment]
This should be native functionality implemented for each supported protocol. Probably https://github.com/minio/minio-go could simplify the integration.

**Comment by MaxDiOrio** [comment]
Yes...native is probably the way to go.

**Comment by Serrvosky** [comment]
That solution could be awesome and a massive improve to Victoria Metrics. Can bring a must have feature with the simple approach like Victoria Metrics got us used.

It would be nice to have a flag like `timeToSnapshot`, where Vitoria Metric's storage nodes can read an interval of time to make a snapshots and send it to the cloud, leaving nothing more that a proper logs .
If something happen there are a lot of log's aggregation and processing tools (like datadog, elastiksearch, loki+grafana) where users can check error logs.

**Comment by valyala** [comment]
FYI, there is [vmbackup](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/app/vmbackup/README.md) tool for full and incremental backups of VictoriaMetrics data on S3 and GCS. The tool will be included into the next VictoriaMetrics release. Now it can be built from sources according to [these instructions](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/app/vmbackup/README.md#how-to-build-from-sources).

> It would be nice to have a flag like timeToSnapshot, where Vitoria Metric's storage nodes can read an interval of time to make a snapshots and send it to the cloud

It should be easy to write a small script around `vmbackup` tool, which would perform this. See [this use case](https://github.com/VictoriaMetrics/VictoriaMetrics/blob/master/app/vmbackup/README.md#smart-backups).

**Comment by perriea** [comment]
Very cool ! We will be able to put a cron job in Kubernetes to backup the TSDB 👍

**Comment by valyala** [comment]
Published an article about `vmbackup` tool -  https://medium.com/@valyala/speeding-up-backups-for-big-time-series-databases-533c1a927883

**Comment by valyala** [comment]
https://objectivefs.com/ looks promising for transparently storing VictoriaMetrics data on object storage  such as S3, GCS or various s3-compatible on-prem systems.

**Comment by faceair** [comment]
https://github.com/juicedata/juicefs juicefs is similar to objectivefs

**Comment by sreeram-boyapati** [comment]
I suggest integration with Apache Hudi for streaming updates to a data lake. Can solve the backup problem.

**Comment by Xtigyro** [comment]
@valyala Any update on this feature request?

**Comment by abelal83** [comment]
> https://github.com/juicedata/juicefs juicefs is similar to objectivefs
> 
> 
> 
> **update:**
> 
> I tested juicefs for poor random read performance and it should not work as a storage for VictoriaMetrics.

For clarity, are you suggesting juicefs should not be used with Victoria Metrics?

**Comment by pythiankerr** [comment]
One cautionary thing about S3: there is rate-limiting on a per-prefix ('folder') level, so you'd need to be careful about how you represented folder structure.

One of the attractions of using S3 is that it's cheaper and has lifecycle support; there'd be some similar experience with Elasticsearch's (commercial) support for S3 in its 'frozen' storage tier. For example, reading data from such a tier should perhaps be an explicit opt-in (to set user expectations and facilitate an asynchronous interface), and potentially access-controlled based on role to protect against cost overruns.

But probably the biggest 'hidden' risk to trying to shim object storage into the virtual file-system is the additional latency of open and closing files. Take NFS (or Amazon EFS) for example, where 'open' and 'close' are synchronous operations, and much much slower than you might imagine (eg. low hundreds of such operations per second) which is what makes NFS less suited for small-file operations. Linux read-ahead could also be a concern for random-access IO; again taking NFS as an example, the default readahead for NFS is 15MB (which is adjustable). This is a very large read-ahead and could end up incurring a lot more traffic if you only need a few KB of random-access data.

To be clear, I'm not suggesting NFS as a supported storage for VictoriaMetrics, just pointing out the similarities of other stacked-filesystem drivers.

Supporting NFS or S3 as an archive tier in a similar way to Elasticsearch's Frozen Tier would be useful. VM might then pull back, load, query and then age out again. The biggest challenge I think would be how the user would signal their intent that data should be come back through a very slow, preferably asynchronous, query path.

Hope that's useful,
Cameron

**Comment by pachico** [comment]
I would love to see a feature that distinguishes between HOT and COLD data, where COLD is older than X and can be stored in S3 or similar.
This is what ClickHouse, for instance, can do.

I assume by this that users (such as myself) are willing to sacrifice query response time that includes COLD data in favour of not having to deal with retention anymore.

**Comment by MaxDiOrio** [comment]
I ended up moving from VM to Mimir for metrics storage mainly because of the storage issue. With Mimir and the other Grafana projects I can use S3 and primary storage. Since my Mimir instance is also in AWD I get 10Gbps low latency accesss to S3 storage.

By not allowing object sotorage, for primary and cold storage, I can’t use VM as other storage options are far more expensive.

**Comment by pachico** [comment]
I'm afraid I will also have to move from VM to mimir precisely because of the storage limitation.
A hot in disk, cold in S3 solution would have been ideal.

**Comment by p4ranoidandro1d** [comment]
Hi guys,

We are moving some k8s clusters to Azure, and there are limitations regarding the number of pv you can assign to a host.
I gave it a try using blobfuse driver but vmstorage crashes quite a lot.

`
unexpected fault address 0x7f7a38c740ac
fatal error: fault
[signal SIGBUS: bus error code=0x2 addr=0x7f7a38c740ac pc=0x46ce09]

goroutine 963 [running]:
runtime.throw({0x9007ec?, 0xc00009cc78?})
        runtime/panic.go:1047 +0x5d fp=0xc000677318 sp=0xc0006772e8 pc=0x4381fd
runtime.sigpanic()
        runtime/signal_unix.go:832 +0x125 fp=0xc000677368 sp=0xc000677318 pc=0x44e9a5
runtime.memmove()
        runtime/memmove_amd64.s:339 +0x3c9 fp=0xc000677370 sp=0xc000677368 pc=0x46ce09
github.com/VictoriaMetrics/VictoriaMetrics/lib/fs.(*ReaderAt).MustReadAt(0xc000ad2660, {0xc00035aa00, 0x12c, 0xc000f731d0?}, 0x0)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/fs/reader_at.go:57 +0x21a fp=0xc000677490 sp=0xc000677370 pc=0x691f9a
github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset.(*partSearch).readIndexBlock(0xc000183380, 0xc00016bc20)
        github.com/VictoriaMetrics/VictoriaMetrics/lib/mergeset/part_search.go:281 +0x105 fp=0xc000677550 sp=0xc000677490 pc=0x6d79a5
`

i dont if it's related to the use of a blob backend storage (https://learn.microsoft.com/en-us/azure/storage/blobs/blobfuse2-what-is#differences-between-the-linux-file-system-and-blobfuse2) but that's being a problem.

Just to let you know one more use case. A move to a s3/blob like compatible solution is whats next for us.

**Comment by pachico** [comment]
We are migrating to Mimir precisely because of the lack of a cold object storage.
I wish we could keep using VM but volumes are expensive in AWS.
I am willing to sacrifice latency when querying old data to mitigate costs.

**Comment by wasilak** [comment]
Same here: after doing research and couple of POCs my company is willing to drop VM and move to Mimir, even though costs seem larger and cluster setup more complex, but this brings benefits of long term location agnostic object storage.

We are growing and cannot afford inflexibility when using EBS volumes associated with single availability zone (in AWS) because it lacks the ability to easily move data across VM clustered storage instances spread across zones/regions. It Is not strictly related to object storage, but having it, we would be able to limit VM storage retention to e.g. 1h and be way more flexible when it comes to moving instances around. Currently in order to preserve data we need to stop writes (keep reads, though) to particular VM instance and wait until the end of data retention period, which is 90 days... Or have downtime and manually use `vmctl`.

**Comment by pachico** [comment]
In case it helps, ClickHouse solves precisely this use case with multiple tier storage. Your hot (frequent and fast) data sits in your solid state storage and your cold (infrequent and slow) data sits in S3 or compatible.

**Comment by faceair** [comment]
Agree with @pachico , maybe we can store the big part in the object store, the indexdb and small part on local disk.

**Comment by valyala** [comment]
> I wish we could keep using VM but volumes are expensive in AWS.

VictoriaMetrics [works perfectly on HDD-based disks](https://victoriametrics.com/blog/benchmark-100m/) such as `st1` disks. They [cost $45 TB/month](https://aws.amazon.com/ebs/pricing/). S3 storage [costs $23 TB/month](https://aws.amazon.com/s3/pricing/). While that's twice more expensive than `st1` disks, these disks provide the following benefits on top of S3 storage:
- **Much lower** IO latency (e.g. milliseconds for `st1` disks vs hundreds of milliseconds for S3 storage). This results in higher query performance, especially when the query needs to read data from big number of time series.
- **Much lower** possibility of IO errors. This translates directly to **much lower** IO-induced error rate during queries.
- Zero costs for data access. S3 storage may become **very** expensive during intensive queries - see the [data transfer pricing and pricing for requests & data retrieval](https://aws.amazon.com/s3/pricing/). These costs can quickly outgrow storage costs.

> We are migrating to Mimir precisely because of the lack of a cold object storage.

Storage costs for Mimir are usually higher than storage costs for VictoriaMetrics, since Mimir has not so good compression rate for the stored data. See [this article](https://victoriametrics.com/blog/mimir-benchmark/) for details.

P.S.: We do not abandon plans for implementing S3-based storage for VictoriaMetrics. The recently added commit 43b24164efdd35c2641a289ab93f5cb94279cb60 removes one of the biggest stumbling blocks on the road to S3-based storage support in VictoriaMetrics: [S3 doesn't support efficient file renaming](https://stackoverflow.com/questions/39469698/rename-folder-in-amazon-s3-slow-can-it-be-resolved-using-the-following), while VictoriaMetrics was using part folder renaming during [background merge](https://docs.victoriametrics.com/#storage).

**Comment by syepes** [comment]
@valyala When implementing this feature it would be great to also add support for swift its mostly compatible with the S3 API: https://docs.openstack.org/swift/stein/s3_compat.html

This removes the S3 IO/Networking costs if it's hosted locally :-)

**Comment by leandropadua** [comment]
We are looking forward for this feature too. 
I believe this has a potential to remove the need for backups.

**Comment by elops-od** [comment]
We are looking forward for this feature too.

**Comment by jicki** [comment]
This seems to be a widespread need

**Comment by ak-aozasa** [comment]
We are looking forward for this feature too!!

**Comment by chukynax** [comment]
We are looking forward for this feature too.

**Comment by butangero** [comment]
Ditto

**Comment by EsDmitrii** [comment]
Awesome feature, we are looking forward for this feature too!

**Comment by dili-pk** [comment]
Any update on this?
Does VM have a S3 integration support as of today? 
I do not see anything about it in the document.. Or is this still in the roadmap?

**Comment by hagen1778** [comment]
> Does VM have a S3 integration support as of today?

VM doesn't support this feature yet. We don't have it listed in our backlog.
We still this [comment](https://github.com/VictoriaMetrics/VictoriaMetrics/issues/38#issuecomment-1484006165) stands.

**Comment by zer0def** [comment]
Linked comment provides a highly AWS-centric point of view, where it might be missing the forest for the trees, as there are numerous more affordable providers on the market, aside from the multitude of solutions exposing S3 interfaces.

With that in mind, one has to grant that there's the possibility of mounting an S3 bucket using rclone or s3fs, even if it also communicates overly-tightened coupling with filesystem-like interfaces.

**Comment by rorsten** [comment]
> Linked comment provides a highly AWS-centric point of view, where it might be missing the forest for the trees, as there are numerous more affordable providers on the market, aside from the multitude of solutions exposing S3 interfaces.
> 
> With that in mind, one has to grant that there's the possibility of mounting an S3 bucket using rclone or s3fs, even if it also communicates overly-tightened coupling with filesystem-like interfaces.

`pd-standard` persistent disk on GCP; I think `Standard_LRS` for azure.

**Comment by davinkevin** [comment]
This one is the only reason for VM to be excluded from an evaluation I did for some companies 😓. 

From a design point of view, it makes sense, but offering a tool able to be "stateless" with data externalised to object storage is something very high in operators wish-list (when grafana, prometheus & others have it).

**Comment by hagen1778** [comment]
@davinkevin could you explain more about being "stateless"?

* Prometheus doesn't support object storage
* Thanos requires stateful Prometheus to run and process queries for recent data
* Mimir requires Ingesters to store last 24h of data (by default, AFAIK) on local filesystem to improve query speed for recent data

Since all 3 solutions share the same code base, all are stateful: they need to store 2h of data in memory before persisting a data part on disk. Only after this data part can be uploaded to the object storage.

**Comment by davinkevin** [comment]
The most recent case I work on is a bit specific (support-bundle, data extraction for by a product), which makes `prom`+`thanos`, `mimir` or just `vector`+`object storage` valid solutions. 

We don't need scalability and all important stuff done by victoria (or any other tool), because we have a very low number of applications doing push.

However, about others cases, when people wants to store 1, 2 or x years of data, object storage is way better from an operation perspective. 

Stateless is maybe not the right word to describe it, but means "the *important* state is not in the cluster in a PVC I have to manage" but just in buckets and all state(full) things you mentioned are implementation details, but still simpler to manage than "the full state in the cluster". 

Can you elaborate downside you have in mind about Object Storage in this context?

**Comment by hagen1778** [comment]
> Can you elaborate downside you have in mind about Object Storage in this context?

In the context of just having a single source of truth somewhere in the cloud instead of it sitting in PVC having Object Storage integration is beneficial, I agree.
But the common downsides of having telemetry data stored in the ObjectStorage are listed here https://github.com/VictoriaMetrics/VictoriaMetrics/issues/38#issuecomment-1484006165

**Comment by kappa-wingman** [comment]
https://cloud.google.com/learn/what-is-object-storage

Object storage is low cost and highly scalable. Resiliency is also great.
It may be costly or not easy for the case of very large block storage.
I look forward to see VM supporting object storage.

**Comment by zer0def** [comment]
Clearly, finger-pointing to a response doesn't seem to achieve the expected clarifying result, considering currently ongoing boomeranging.

Perhaps it might be more enlightening to analyze `lib/storage` and `lib/encoding`, extensively used by `vmstorage` to perform it's functionality.

**Comment by sanchpet** [comment]
Looking forward for this feature

**Comment by justinrush** [comment]
We migrated from Cortex to VM so I understand concerns of increased complexity with having something that syncs to/from an external storage provider. That said, there are three large operational issues we have to overcome to operate VM at scale (i.e. one siloed instance in each Kubernetes deployment  (regulations prevent aggregation)... 400+ deployments).

1. Decoupled storage and compute. Azure doesn't make it super easy to create and manage PVCs outside of Kubernetes clusters at scale. We need to be able to wipe out an AKS instance and deploy another one and have all the apps be able to access the same storage. This helps us work around things like regional outages or odd ARM instances where AKS can get into a wonky state.
2. Avoid the need for quarterly growth tasks. Azure bills PVCs based on provisioned size (e.g. 4g, 16g, 32g, ... 1t, 2t, 4t, etc). If you provision 4TB PVCs because you know you'll grow into it, you get to pay for 4TB for the entire time, even when you only use 1G. To work around this we have a partially scripted way to grow the PVCs used by the vmstorage stateful set once per quarter to account for growth.
3. Related to 2, you pay for provisioned storage in Azure, not what you use. This can make it more expensive. The devil is in the details on this one due to a higher transactional cost for Azure blob versus disk operations on PVCs.

The lack of object storage isn't holding us back for our one large instance that aggregates non-regulated data (e.g. pod resource stats, etc). The operational challenges on a single instance are very manageable. We have concerns when we try and think about running 400+ cluster instances and Azure blob addresses those concerns.

It may be a nice way to crawl/walk/run by targeting object storage for logs first and then extending support to normal TSDB metrics. That would at least allow us to move off of Loki in our 400+ instances we have today.

**Comment by ZVilusinsky** [comment]
Frankly, the lack of object storage backend is what makes us consider migrating from VM to a solution for long term storage that does support it.

**Comment by alen-z** [comment]
I can offer to talk more about RelayBoost. This is an umbrella for me saying: MinIO with object store backend combined with JuiceFS that exposes the MinIO storage as Kubernetes resources (PVCs, PVs). You attach any application to it, with PVC in RWX mode as benefit, if/when needed.

I agree that VM supporting object store as storage backend would be awesome.

**Comment by makasim** [comment]
This feature will be considered after it is implemented in VictoriaLogs (see https://github.com/VictoriaMetrics/VictoriaLogs/issues/48). It might reuse some libraries from there. Please also note that any Object Storage implementation is blocked by https://github.com/VictoriaMetrics/VictoriaMetrics/issues/7599. It has to be implemented first.
